R/cube-transforms.R

```diff
@@ -0,0 +1,177 @@
+#' combination of these can be given, by default all will be given.
+#'
+#' `subtotalArray(cube)` is a convenience function that is equivalent to
+#' `applyTransforms(cube, include = c("subtotals"))`
+#'
+#' @param x a CrunchCube
+#' @param ary an array to use, if not using the default array from the cube
+#' itself. (Default: not used, pulls an array from the cube directly)
+#' @param ... arguments to pass to `calcTransforms`, for example `include`
+#' @param headings for `subtotalArray`: a logical indicating if the headings
+#' should be included with the subtotals (default: `FALSE`)
+#'
+#' @return an array with any transformations applied
+#'
+#' @examples
+#' \dontrun{
+#' transformed_array <- applyTransforms(crtabs(~opinion, ds))
+#'
+#' # to get an array of just the subtotals
+#' subtotalArray(crtabs(~opinion, ds))
+#' #    Agree Disagree 
+#' #       47       35 
+#' 
+#' # to get the full array with the subtotals but not headings
+#' applyTransforms(crtabs(~opinion, ds), include = c("cube_cells", "subtotals"))
+#' #             Strongly Agree             Somewhat Agree                      Agree 
+#' #                         23                         24                         47 
+#' # Neither Agree nor Disagree          Strongly Disagree                   Disagree 
+#' #                         18                         19                         35 
+#' #          Somewhat Disagree 
+#' #                         16 
+#' # to get the full array with the headings but not subtotals
+#' applyTransforms(crtabs(~opinion, ds), include = c("cube_cells", "headings")) 
+#' #               All opinions             Strongly Agree             Somewhat Agree 
+#' #                         NA                         23                         24 
+#' # Neither Agree nor Disagree          Strongly Disagree          Somewhat Disagree 
+#' #                         18                         19                         16 
+#' }
+#'
+#' @aliases subtotalArray
+#' @export
+applyTransforms <- function (x, ary = cubeToArray(x), ...) {
+    # if there are row transforms, calculate them and display them
+    row_trans <- tryCatch(Transforms(data=index(variables(x))[[1]]$view$transform), error = function(e) NULL)
+    if (!is.null(row_trans)) {
+        var_cats <- Categories(data=index(variables(x))[[1]]$categories)
+
+        # TODO: calculate category/element changes
+
+        if (length(dim(ary)) > 1) {
+            off_margins <- seq_along(dim(ary))[-1]
+            dim_names  <- names(dimnames(ary))
+            ary <- apply(ary, off_margins, calcTransforms, row_trans, var_cats, ...)
+            names(dimnames(ary)) <- dim_names
+        } else {
+            ary <- as.array(calcTransforms(ary, row_trans, var_cats, ...))
+        }
+
+        ary <- subsetTransformedCube(ary, x)
+    }
+
+    # TODO: calculate column transforms
+
+
+    return(ary)
+}
+
+subsetTransformedCube <- function (ary, cube) {
```

Comment:
Why does this function need to exist? Is there a way that existing similar functions (`cubeToArray`, `subsetCubeArray`, etc.) can be used?